src/java.base/share/classes/sun/launcher/LauncherHelper.java

```diff
@@ -920,58 +910,65 @@ private static boolean doesExtendFXApplication(Class<?> mainClass) {
         return false;
     }
 
-    private static boolean isStaticMain = false;
-    private static boolean noArgMain = false;
-
-    // Check the existence and signature of main and abort if incorrect.
-    private static void validateMainMethod(Class<?> mainClass) {
+    // Determine the presence of a valid main method on the MainEntry.Builder.mainClass.
+    // abort() if validation fails.
+    private static MainEntry requireValidMainMethod(MainEntry.Builder builder) {
         Method mainMethod = null;
         try {
-            mainMethod = MethodFinder.findMainMethod(mainClass);
-
+            mainMethod = MethodFinder.findMainMethod(builder.mainClass);
             if (mainMethod == null) {
                 // invalid main or not FX application, abort with an error
-                abort(null, "java.launcher.cls.error2", mainClass.getName(),
+                abort(null, "java.launcher.cls.error2", builder.mainClass.getName(),
                       JAVAFX_APPLICATION_CLASS_NAME);
             }
         } catch (Throwable e) {
-            if (mainClass.getModule().isNamed()) {
+            if (builder.mainClass.getModule().isNamed()) {
                 abort(e, "java.launcher.module.error3",
-                      mainClass.getName(), mainClass.getModule().getName(),
+                      builder.mainClass.getName(), builder.mainClass.getModule().getName(),
                       e.getClass().getName(), e.getLocalizedMessage());
             } else {
-                abort(e, "java.launcher.cls.error5", mainClass.getName(),
+                abort(e, "java.launcher.cls.error5", builder.mainClass.getName(),
                       e.getClass().getName(), e.getLocalizedMessage());
             }
         }
+        // main method must be located by now
+        assert mainMethod != null : "main method wasn't located";
 
         int mods = mainMethod.getModifiers();
-        isStaticMain = Modifier.isStatic(mods);
+        boolean isStaticMain = Modifier.isStatic(mods);
         boolean isPublic = Modifier.isPublic(mods);
-        noArgMain = mainMethod.getParameterCount() == 0;
-
+        boolean noArgMain = mainMethod.getParameterCount() == 0;
         if (!PreviewFeatures.isEnabled()) {
             if (!isStaticMain || !isPublic || noArgMain) {
-                  abort(null, "java.launcher.cls.error2", mainClass.getName(),
-                       JAVAFX_APPLICATION_CLASS_NAME);
+                abort(null, "java.launcher.cls.error2", builder.mainClass.getName(),
+                        JAVAFX_APPLICATION_CLASS_NAME);
             }
-            return;
         }
```

Comment:
```suggestion
        if (!PreviewFeatures.isEnabled()) {
            int mods = mainMethod.getModifiers();
            if (!Modifier.isStatic(mods) || !Modifier.isPublic(mods) || mainMethod.getParameterCount() == 0) {
                abort(null, "java.launcher.cls.error2", builder.mainClass.getName(),
                        JAVAFX_APPLICATION_CLASS_NAME);
            }
        }
```
Remove variables that are only used once

Reply:
This can now be removed. Compact source files and instance main methods have been integrated.